### ISSUE TITLE
PPC testsuite failures with Ubuntu 15.10 (GCC 5)

### ISSUE THREAD

**Issue by edelsohn** [JuliaLang/openlibm#120]
```
$ make ARCH=powerpc check

test/test-double
testing double (without inline functions)
Failure: acos (inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: acos (-inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: acos (1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: acos (-1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: asin (inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: asin (-inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: asin (1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: asin (-1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: cos (inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: cos (-inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: sin (inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: sin (-inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: sincos (inf, &sin_res, &cos_res) puts NaN in sin_res plus invalid exception: Exception "Invalid operation" not set
Failure: sincos (-inf, &sin_res, &cos_res) puts NaN in sin_res plus invalid exception: Exception "Invalid operation" not set
Failure: tan (inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: tan (-inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: acosh (-inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: acosh (-1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: atanh (1) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: atanh (-1) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: atanh (1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: atanh (-1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: log (0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log (-0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log (-1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: log10 (0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log10 (-0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log10 (-1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: log1p (-1) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log1p (-2) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: log2 (0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log2 (-0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log2 (-1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: logb (0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: logb (-0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: pow (-0.1, 1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: pow (-0.1, -1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: pow (-10.1, 1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: pow (-10.1, -1.1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: pow (0, -1) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: pow (0, -11) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: pow (-0, -1) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: pow (-0, -11) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: pow (0, -2) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: pow (0, -11.1) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: pow (-0, -2) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: pow (-0, -11.1) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: sqrt (-1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: sqrt (-inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: gamma (0) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: gamma (-3) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: lgamma (0) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: lgamma (-3) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: tgamma (0) == inf plus divide-by-zero: Exception "Divide by zero" not set
Failure: tgamma (-0) == inf plus divide-by-zero: Exception "Divide by zero" not set
Failure: tgamma (-2) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: tgamma (-inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: Test: rint (0.5) == 0.0
Result:
 is:          1.00000000000000000000e+00   0x1.00000000000000000000p+0
 should be:   0.00000000000000000000e+00   0x0.00000000000000000000p+0
 difference:  1.00000000000000000000e+00   0x1.00000000000000000000p+0
 ulp       :  4503599627370496.0000
 max.ulp   :  0.0000
Failure: Test: rint (2.5) == 2.0
Result:
 is:          3.00000000000000000000e+00   0x1.80000000000000000000p+1
 should be:   2.00000000000000000000e+00   0x1.00000000000000000000p+1
 difference:  1.00000000000000000000e+00   0x1.00000000000000000000p+0
 ulp       :  2251799813685248.0000
 max.ulp   :  0.0000
Failure: Test: rint (4.5) == 4.0
Result:
 is:          5.00000000000000000000e+00   0x1.40000000000000000000p+2
 should be:   4.00000000000000000000e+00   0x1.00000000000000000000p+2
 difference:  1.00000000000000000000e+00   0x1.00000000000000000000p+0
 ulp       :  1125899906842624.0000
 max.ulp   :  0.0000
Failure: Test: rint (-1.5) == -2.0
Result:
 is:         -1.00000000000000000000e+00  -0x1.00000000000000000000p+0
 should be:  -2.00000000000000000000e+00  -0x1.00000000000000000000p+1
 difference:  1.00000000000000000000e+00   0x1.00000000000000000000p+0
 ulp       :  2251799813685248.0000
 max.ulp   :  0.0000
Failure: Test: rint (-3.5) == -4.0
Result:
 is:         -3.00000000000000000000e+00  -0x1.80000000000000000000p+1
 should be:  -4.00000000000000000000e+00  -0x1.00000000000000000000p+2
 difference:  1.00000000000000000000e+00   0x1.00000000000000000000p+0
 ulp       :  1125899906842624.0000
 max.ulp   :  0.0000
Maximal error of `rint'
 is      :  4503599627370496.0000 ulp
 accepted:  0.0000 ulp
Failure: Test: lrint (0.2) == 0
Result:
 is:         1
 should be:  0
Failure: Test: lrint (1.4) == 1
Result:
 is:         2
 should be:  1
Failure: Test: lrint (8388600.3) == 8388600
Result:
 is:         8388601
 should be:  8388600
Failure:Test: llrint (0.2) == 0
Result:
 is:         1
 should be:  0
Failure:Test: llrint (1.4) == 1
Result:
 is:         2
 should be:  1
Failure:Test: llrint (8388600.3) == 8388600
Result:
 is:         8388601
 should be:  8388600
Failure: fmod (inf, 3) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fmod (-inf, 3) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fmod (3, 0) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fmod (3, -0) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: remainder (1, 0) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: remainder (1, -0) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: remainder (inf, 1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: remainder (-inf, 1) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: remquo (1, 0, &x) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: remquo (1, -0, &x) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: remquo (inf, 1, &x) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: remquo (-inf, 1, &x) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fma (inf, 0.0, 1.0) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fma (-inf, 0.0, 1.0) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fma (0.0, inf, 1.0) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fma (0.0, -inf, 1.0) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fma (inf, inf, -inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fma (-inf, inf, inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fma (inf, -inf, inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: fma (-inf, -inf, -inf) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: y0 (-1.0) == NaN: Exception "Invalid operation" not set
Failure: y0 (0.0) == -inf: Exception "Divide by zero" not set
Failure: y1 (-1.0) == NaN: Exception "Invalid operation" not set
Failure: y1 (0.0) == -inf: Exception "Divide by zero" not set
Failure: yn (0, -1.0) == NaN: Exception "Invalid operation" not set
Failure: yn (0, 0.0) == -inf: Exception "Divide by zero" not set
Failure: yn (1, -1.0) == NaN: Exception "Invalid operation" not set
Failure: yn (1, 0.0) == -inf: Exception "Divide by zero" not set

Test suite completed:
  1118 test cases plus 932 tests for exception flags executed.
  97 errors occurred.
```

**Comment by ViralBShah** [comment]
What happens with gcc-4.8?

**Comment by simonbyrne** [comment]
I just installed gcc 5.3 on my (big endian) test instance, and all tests pass. Are you using little endian?

**Comment by edelsohn** [comment]
Yes this is all little endian.  All of this work should have been little endian.

**Comment by simonbyrne** [comment]
Okay, I've managed to get a little endian instance up and running, and I see the same on gcc 4.8. I'll see what I can do.

**Comment by simonbyrne** [comment]
If you compile it to use host `fenv.h`,
```
make ARCH=powerpc CFLAGS='-DOPENLIBM_USE_HOST_FENV_H' test
```
then the `rint` errors go away, but most of the other errors remain.

I'm guessing this is an endianness issue in the inline assembly used for interacting with the control register.

**Comment by edelsohn** [comment]
I just opened FreeBSD PR 207529 to track this as well and talking with the FreeBSD team.  No reason to fix this multiple times.

The struct

union __fpscr {
        double __d;
        struct {
                __uint32_t __junk;
                fenv_t __reg;
        } __bits;
};

is not correct for little endian.

**Comment by yuyichao** [comment]
Isn't that exactly the fix in #121?
